From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ShyAlly <1447610+ShyAlly@users.noreply.github.com>
Date: Sat, 10 Jan 2026 12:51:39 -0500
Subject: [PATCH] LHC - Allowing falling block entity duplication


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 8367e5cdf545e0434cf6fdcca9c09a14a5dbd7e7..ae91ec2b4703ccfdebd2e2ad22506145296fbd7d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -92,6 +92,7 @@ import net.minecraft.world.Nameable;
 import net.minecraft.world.attribute.EnvironmentAttributes;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.damagesource.DamageSources;
+import net.minecraft.world.entity.item.FallingBlockEntity;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.Projectile;
@@ -3516,6 +3517,12 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         if (this.level() instanceof ServerLevel serverLevel) {
             this.processPortalCooldown();
             if (this.portalProcess != null) {
+                // LHC start - allow duplicating falling blocks
+                boolean canUsePortal = this.canUsePortal(false);
+                if (!canUsePortal && !isAlive() && this instanceof FallingBlockEntity fbe) {
+                    fbe.spawnAtLocation(serverLevel, fbe.blockState.getBlock());
+                }
+                // LHC end
                 if (this.portalProcess.processPortalTeleportation(serverLevel, this, this.canUsePortal(false))) {
                     ProfilerFiller profilerFiller = Profiler.get();
                     profilerFiller.push("portal");
