From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ShyAlly <1447610+ShyAlly@users.noreply.github.com>
Date: Fri, 12 Dec 2025 09:44:26 -0500
Subject: [PATCH] LHC - Adjusted angle calculation


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1126c1c23206e7d68752a5ccf4194461851bb119..3f8e984d97ec2785cab690cbbbad8672e3a683d0 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4757,7 +4757,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         }
 
                         float yRot;
-                        if (finalPosition.getX() == 0 && finalPosition.getZ() == 0) {
+                        if (finalPosition.getX() == 0 || finalPosition.getZ() == 0) {
                             yRot = Direction.WEST.toYRot();
                         } else {
                             double degrees = Math.toDegrees(Math.atan2(finalPosition.getZ() * -1, finalPosition.getX() * -1));
@@ -4765,6 +4765,9 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                             if (yRot > 180) {
                                 yRot -= 360;
                             }
+                            if (yRot < -180) {
+                                yRot += 360;
+                            }
                         }
                         LOGGER.info("Calculated yRot: {}", yRot);
 
