From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ShyAlly <1447610+ShyAlly@users.noreply.github.com>
Date: Fri, 12 Dec 2025 11:05:48 -0500
Subject: [PATCH] LHC - Renamed some variables, redid some logic


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 311bad1665fe65049fa642396de3aff0643422ae..58b426b7f708348db5877daf49f5f049d86ce1fd 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4735,20 +4735,21 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         // If it fails we'll end up at the regular end spawn
                         BlockPos finalPosition = null;
                         for (int i = 0; i < radius * 4; i++) {
-                            BlockPos potentialPosition = new BlockPos(center.getX() + random.nextInt(radius * 2) - radius, center.getY(), center.getZ() + random.nextInt(radius * 2) - radius);
-                            BlockPos adjustedSpawn = destination.getHeightmapPos(Heightmap.Types.WORLD_SURFACE, potentialPosition);
-                            if (adjustedSpawn.getY() > 0) {
-                                BlockState landingBlock = destination.getBlockStateIfLoaded(adjustedSpawn);
+                            BlockPos randomPositionYZero = new BlockPos(center.getX() + random.nextInt(radius * 2) - radius, center.getY(), center.getZ() + random.nextInt(radius * 2) - radius);
+                            BlockPos potentialPosition = destination.getHeightmapPos(Heightmap.Types.WORLD_SURFACE, randomPositionYZero);
+                            if (potentialPosition.getY() > 0) {
+                                BlockState landingBlock = destination.getBlockStateIfLoaded(potentialPosition);
                                 if (landingBlock == null) {
                                     LOGGER.info("Random end spawn - Potential location wasn't loaded?");
                                     continue;
                                 }
+                                LOGGER.info("Random end spawn - Potential {} is {}", potentialPosition, landingBlock);
                                 if (!landingBlock.isSolid()) {
                                     continue;
                                 }
 
                                 // Yay we found one
-                                finalPosition = adjustedSpawn;
+                                finalPosition = potentialPosition;
                                 break;
                             }
                         }
